ext/File-Find/t/taint.t: intermittent failure, possibly due to stale file handle

Recent work done on the tests in `ext/File-Find/` may have introduced intermittent test failures in `ext/File-Find/t/*.t`.  These test failures are appearing in test environments which are, I suspect, resource-challenged.  One of them appears to be a Raspberry Pi running Linux; the other is my OpenBSD-6.9 VM.

**Linux**

https://perl5.test-smoke.org/report/5029413

This is one of George Greer's smokers, which was testing one of @khwilliamson's branches (now superseded).

```
Automated smoke report for branch smoke-me/khw-env 5.37.8 patch d26d6c57c7f3c558bf9b5e29bab7752f9067c18f 
v5.37.8-152-gd26d6c57c7.
develpi.local: ARMv7 Processor rev 3 (v7l) (300 MHz) (armv7l/4 cpus)
    on        linux 5.10.103-v7l+ [Raspbian GNU/Linux 10.13]
    using     ccache gcc version 8.3.0 (*1)
...

Test failures:
  ../ext/File-Find/t/taint.t (Non-zero exit status: 116 Bad plan.  You planned 45 tests but ran 1.)
      [stdio] -Dcc="ccache gcc"
```

In the [log file](https://m-l.org/~perl/smoke/perl/arm-linux-gnueabihf-thread-multi-64int/smoke-me_gcc_quick/Karl%20Williamson/logd26d6c57c7f3c558bf9b5e29bab7752f9067c18f.log.gz) which George provides we see this:
```
Can't cd to .. from ./FF_find_t_9kHmlb/for_find/fb/fbc: Stale file handle at /home/perl/smoke/perl/s
moke-me_gcc_quick/build/ext/File-Find/../../lib/File/Find.pm line 477.
# Looks like your test exited with 116 just after 1.
# locales available: C C.UTF-8 POSIX en_US.utf8
[2023-01-20 18:20:01-0500] [died test] ../ext/File-Find/t/taint.t..................................
[2023-01-20 18:20:01-0500] [TAP-error test] ../ext/File-Find/t/taint.t..............................
....
make: *** [makefile:865: test_harness] Error 1

error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/b
in/../lib/Test/Smoke.pm line 195.
[2023-01-20 18:20:02-0500] 
[2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-20 18:20:02-0500]     Non-zero exit status: 116
[2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-20 18:20:02-0500]     Bad plan.  You planned 45 tests but ran 1.
```
Note that mention of a **Stale file handle**

**OpenBSD-6.9**

I've observed failures in `taint.t` twice in recent weeks in this VM:  once while doing a regular smoke-test run, once while testing the perl-5.37.8 tarball earlier this week.  The reports are similar, so let's present the smoke-test run.

https://perl5.test-smoke.org/report/5029144

```
Automated smoke report for branch blead 5.37.8 patch bc01770b32f9d9bf9d313d2eec370c41822cd61c 
v5.37.7-146-gbc01770b32.
perl-reporter-11: Intel(R) Xeon(R) CPU E31220 @ 3.10GHz (3093 MHz) (amd64/2 cpus)
    on        openbsd 6.9
    using     cc version OpenBSD Clang 10.0.1 (*1)
...
Test failures:
  ../ext/File-Find/t/taint.t (Non-zero exit status: 2 Bad plan.  You planned 45 tests but ran 0.)
      [stdio] -Duseithreads -Accflags="-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -fno-delete-null-pointer-checks"
```
From the log file:
```
Can't cd to .. from ./FF_find_t_uHzKN6/for_find/fc/fca: No such file or directory at /home/jkeenan/p
5smoke/perl-current/ext/File-Find/../../lib/File/Find.pm line 477.
# Looks like your test exited with 2 before it could output anything.
[2023-01-16 01:34:06+0000] [died test] ../ext/File-Find/t/taint.t..................................
[2023-01-16 01:34:06+0000] [TAP-error test] ../ext/File-Find/t/taint.t..............................
....
*** Error 1 in /home/jkeenan/p5smoke/perl-current (makefile:851 'test_harness')

error while running harness target 'test_harness': 2 at /home/jkeenan/p5smoke/lib/Test/Smoke/App/RunSmoke.pm line 118.
[2023-01-16 01:34:07+0000] 
[2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-16 01:34:07+0000]     Non-zero exit status: 2
[2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
[2023-01-16 01:34:07+0000]     Bad plan.  You planned 45 tests but ran 0.
```
This is similar to George's report except that it lacks the system warning about "Stale file handle".

This is the first time I've ever heard of a stale file handle.  DDG search turned up some interesting articles, but this is beyond my expertise.

@demerphq, the revisions you did in `ext/File-Find` earlier this month certainly look plausible, but could you take a look to see if they might have generated these failures?

Thank you very much.

On Mon, 23 Jan 2023, 04:21 James E Keenan, ***@***.***> wrote:

> Recent work done on the tests in ext/File-Find/ may have introduced
> intermittent test failures in ext/File-Find/t/*.t. These test failures
> are appearing in test environments which are, I suspect,
> resource-challenged.
>
Are the tests being run in parallel or in series?

> One of them appears to be a Raspberry Pi running Linux; the other is my
> OpenBSD-6.9 VM.
>
> *Linux*
>
> https://perl5.test-smoke.org/report/5029413
>
> This is one of George Greer's smokers, which was testing one of
> @khwilliamson <https://github.com/khwilliamson>'s branches (now
> superseded).
>
> Automated smoke report for branch smoke-me/khw-env 5.37.8 patch d26d6c57c7f3c558bf9b5e29bab7752f9067c18f
> v5.37.8-152-gd26d6c57c7.
> develpi.local: ARMv7 Processor rev 3 (v7l) (300 MHz) (armv7l/4 cpus)
>     on        linux 5.10.103-v7l+ [Raspbian GNU/Linux 10.13]
>     using     ccache gcc version 8.3.0 (*1)
> ...
>
> Test failures:
>   ../ext/File-Find/t/taint.t (Non-zero exit status: 116 Bad plan.  You planned 45 tests but ran 1.)
>       [stdio] -Dcc="ccache gcc"
>
> In the log file
> <https://m-l.org/~perl/smoke/perl/arm-linux-gnueabihf-thread-multi-64int/smoke-me_gcc_quick/Karl%20Williamson/logd26d6c57c7f3c558bf9b5e29bab7752f9067c18f.log.gz>
> which George provides we see this:
>
> Can't cd to .. from ./FF_find_t_9kHmlb/for_find/fb/fbc: Stale file handle at /home/perl/smoke/perl/s
> moke-me_gcc_quick/build/ext/File-Find/../../lib/File/Find.pm line 477.
> # Looks like your test exited with 116 just after 1.
> # locales available: C C.UTF-8 POSIX en_US.utf8
> [2023-01-20 18:20:01-0500] [died test] ../ext/File-Find/t/taint.t..................................
> [2023-01-20 18:20:01-0500] [TAP-error test] ../ext/File-Find/t/taint.t..............................
> ....
> make: *** [makefile:865: test_harness] Error 1
>
> error while running harness target 'test_harness': 2 at /home/perl/smoke/ts/Test-Smoke-1.72-custom/b
> in/../lib/Test/Smoke.pm line 195.
> [2023-01-20 18:20:02-0500]
> [2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-20 18:20:02-0500]     Non-zero exit status: 116
> [2023-01-20 18:20:02-0500] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-20 18:20:02-0500]     Bad plan.  You planned 45 tests but ran 1.
>
> Note that mention of a *Stale file handle*
>
> *OpenBSD-6.9*
>
> I've observed failures in taint.t twice in recent weeks in this VM: once
> while doing a regular smoke-test run, once while testing the perl-5.37.8
> tarball earlier this week. The reports are similar, so let's present the
> smoke-test run.
>
> https://perl5.test-smoke.org/report/5029144
>
> Automated smoke report for branch blead 5.37.8 patch bc01770b32f9d9bf9d313d2eec370c41822cd61c
> v5.37.7-146-gbc01770b32.
> perl-reporter-11: Intel(R) Xeon(R) CPU E31220 @ 3.10GHz (3093 MHz) (amd64/2 cpus)
>     on        openbsd 6.9
>     using     cc version OpenBSD Clang 10.0.1 (*1)
> ...
> Test failures:
>   ../ext/File-Find/t/taint.t (Non-zero exit status: 2 Bad plan.  You planned 45 tests but ran 0.)
>       [stdio] -Duseithreads -Accflags="-DNO_LOCALE_NUMERIC -DNO_LOCALE_COLLATE -fno-delete-null-pointer-checks"
>
> From the log file:
>
> Can't cd to .. from ./FF_find_t_uHzKN6/for_find/fc/fca: No such file or directory at /home/jkeenan/p
> 5smoke/perl-current/ext/File-Find/../../lib/File/Find.pm line 477.
> # Looks like your test exited with 2 before it could output anything.
> [2023-01-16 01:34:06+0000] [died test] ../ext/File-Find/t/taint.t..................................
> [2023-01-16 01:34:06+0000] [TAP-error test] ../ext/File-Find/t/taint.t..............................
> ....
> *** Error 1 in /home/jkeenan/p5smoke/perl-current (makefile:851 'test_harness')
>
> error while running harness target 'test_harness': 2 at /home/jkeenan/p5smoke/lib/Test/Smoke/App/RunSmoke.pm line 118.
> [2023-01-16 01:34:07+0000]
> [2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-16 01:34:07+0000]     Non-zero exit status: 2
> [2023-01-16 01:34:07+0000] ../ext/File-Find/t/taint.t..................................FAILED
> [2023-01-16 01:34:07+0000]     Bad plan.  You planned 45 tests but ran 0.
>
> Why does it say the test failed twice?  Did it actually launch the test
twice?

This is similar to George's report except that it lacks the system warning
> about "Stale file handle".
>
> This is the first time I've ever heard of a stale file handle. DDG search
> turned up some interesting articles, but this is beyond my expertise.
>
> @demerphq <https://github.com/demerphq>, the revisions you did in
> ext/File-Find earlier this month certainly look plausible, but could you
> take a look to see if they might have generated these failures
>

I'll take a look, but off the top of my head I can't see how this patch
would trigger an error like this.

The duplicate fail report in that test suggests something else is at play.
Was this test run under 'test_harness' or 'test'?

Do smokers set anything that might derandomize the tests?

Could the smoke framework contain a big and run the tests twice?

If the answer to those last two questions is yes then these failures would
be quite understandable.

I think we need to make harness report how many jobs it was running when it
finishes.

Yves

>

--- a/ext/File-Find/t/find.t
+++ b/ext/File-Find/t/find.t
@@ -60,8 +60,6 @@ my $orig_dir = cwd();
 #     };
 # }
 
-cleanup();
-
 ##### Sanity checks #####
 # Do find() and finddepth() work correctly with an empty list of
 # directories?
@@ -89,8 +87,16 @@ is($::count_taint, 1, "'finddepth' found exactly 1 file named 'taint.t'");
 
 my $FastFileTests_OK = 0;
 
+my $test_root_dir = cwd();
+my $test_temp_dir = tempdir("FF_find_t_XXXXXX",CLEANUP=>1);
+chdir($test_temp_dir) or die "Failed to chdir to '$test_temp_dir': $!";
+
 sub cleanup {
-    chdir($orig_dir);
+    # doing this in two steps avoids the need to know about
+    # directory separators, which is helpful as we override
+    # the File::Spec heirarchy, so we can't ask it to help us here.
+    chdir($test_root_dir) or die "Failed to chdir to '$test_root_dir': $!";
+    chdir($test_temp_dir) or die "Failed to chdir to '$test_temp_dir': $!";
     my $need_updir = 0;
     if (-d dir_path('for_find')) {
         $need_updir = 1 if chdir(dir_path('for_find'));
@@ -139,6 +145,7 @@ sub cleanup {
     if (-d dir_path('for_find')) {
         rmdir dir_path('for_find') or print "# Can't rmdir for_find: $!\n";
     }
+    chdir($test_root_dir) or die "Failed to chdir to '$test_root_dir': $!";
 }
 
 END {
@@ -236,10 +243,6 @@ sub my_postprocess {
 *file_path_name = \&file_path;
 
 ##### Create directories, files and symlinks used in testing #####
-my $root_dir = cwd();
-my $test_dir = tempdir("FF_find_t_XXXXXX",CLEANUP=>1);
-chdir $test_dir;
-
 mkdir_ok( dir_path('for_find'), 0770 );
 ok( chdir( dir_path('for_find')), "Able to chdir to 'for_find'")
     or die("Unable to chdir to 'for_find'");
@@ -1115,5 +1118,4 @@ if ($^O eq 'MSWin32') {
     like($@, qr/invalid top directory/,
         "find() correctly died due to undefined top directory");
 }
-chdir $root_dir; # let File::Temp cleanup - Switch to `defer {}` one day
 done_testing();

--- a/ext/File-Find/t/taint.t
+++ b/ext/File-Find/t/taint.t
@@ -72,11 +72,6 @@ BEGIN {
 
 my $symlink_exists = eval { symlink("",""); 1 };
 
-my $orig_dir = cwd();
-( my $orig_dir_untainted ) = $orig_dir =~ m|^(.+)$|; # untaint it
-
-cleanup();
-
 my $found;
 find({wanted => sub { ++$found if $_ eq 'taint.t' },
                 untaint => 1, untaint_pattern => qr|^(.+)$|}, File::Spec->curdir);
@@ -92,8 +87,22 @@ is($found, 1, 'taint.t found once again');
 my $case = 2;
 my $FastFileTests_OK = 0;
 
+my $test_root_dir; # where we are when this test starts
+my $test_root_dir_tainted = cwd();
+if ($test_root_dir_tainted =~ /^(.*)$/) {
+    $test_root_dir = $1;
+} else {
+    die "Failed to untaint root dir of test";
+}
+my $test_temp_dir = tempdir("FF_taint_t_XXXXXX",CLEANUP=>1);
+chdir($test_temp_dir) or die "Failed to chdir to '$test_temp_dir': $!";
+
 sub cleanup {
-    chdir($orig_dir_untainted);
+    # doing this in two steps avoids the need to know about
+    # directory separators, which is helpful as we override
+    # the File::Spec heirarchy, so we can't ask it to help us here.
+    chdir($test_root_dir) or die "Failed to chdir to '$test_root_dir': $!";
+    chdir($test_temp_dir) or die "Failed to chdir to '$test_temp_dir': $!";
     my $need_updir = 0;
     if (-d dir_path('for_find_taint')) {
         $need_updir = 1 if chdir(dir_path('for_find_taint'));
@@ -120,6 +129,7 @@ sub cleanup {
     if (-d dir_path('for_find_taint')) {
         rmdir dir_path('for_find_taint') or print "# Can't rmdir for_find_taint: $!\n";
     }
+    chdir($test_root_dir) or die "Failed to chdir to '$test_root_dir': $!";
 }
 
 END {
@@ -167,16 +177,6 @@ sub simple_wanted {
 *file_path_name = \&file_path;
 
 ##### Create directories, files and symlinks used in testing #####
-my $root_dir_tainted = cwd();
-my $root_dir;
-if ($root_dir_tainted=~/^(.*)$/) {
-    $root_dir = $1;
-} else {
-    die "Failed to untaint root dir of test";
-}
-my $test_dir = tempdir("FF_taint_t_XXXXXX",CLEANUP=>1);
-chdir $test_dir;
-
 mkdir_ok( dir_path('for_find_taint'), 0770 );
 ok( chdir( dir_path('for_find_taint')), 'successful chdir() to for_find_taint' );
 
@@ -342,4 +342,3 @@ SKIP: {
 
     chdir($cwd_untainted);
 }
-chdir $root_dir; # let File::Temp cleanup - Switch to `defer {}` one day

